bucket_map/src/bucket.rs

```diff
@@ -528,20 +528,21 @@ impl<'b, T: Clone + Copy + PartialEq + std::fmt::Debug + 'static> Bucket<T> {
             // fail early if the data bucket we need doesn't exist - we don't want the index entry partially allocated
             return Err(BucketMapError::DataNoSpace((best_fit_bucket, 0)));
         }
+
         let max_search = self.index.max_search();
         let (elem, elem_ix) = Self::find_index_entry_mut(&self.index, key, self.random)?;
-        let elem = if let Some(elem) = elem {
-            elem
-        } else {
-            let is_resizing = false;
-            self.index.occupy(elem_ix, is_resizing).unwrap();
-            let elem_allocate = IndexEntryPlaceInBucket::new(elem_ix);
-            // These fields will be overwritten after allocation by callers.
-            // Since this part of the mmapped file could have previously been used by someone else, there can be garbage here.
-            elem_allocate.init(&mut self.index, key);
-            elem_allocate
-        };
         if !requires_data_bucket {
+            let elem = if let Some(elem) = elem {
```

Comment:
Note this code is duplicated below. However, the sharing of `&mut` and overlapping lifetimes make extracting this code very difficult due to the borrow checker.